[Seldev] Step15-2,3 완료

p-iknow opened the pull request:
step 15-2와 3(지난번 미션때 적용했습니다.) 완료했습니다. 리뷰 부탁드립니다.

[데모링크](https://p-iknow.github.io/codeSquad_FE/amazone-clone/index.html)

## 진행사항 
- 기존 Carousel 과 병합 및 폴더구조 변경
- 엄격한 Event 처리(mouse, keyboard)
- async 적용 및 제공된 api 활용 
- 피드백 받았던 내용(변수, 함수명)
### 가능 문자열 
 - "iphone8"
 - "bicycle hel"
 - "javascript

### 개선해야 할 것 
- 의도가 담겨 있는 폴더구조(고민중)
- componant 별로 다른 객체 패턴을 가지지 못한 것(carousel은  observer, 자동검색은 mvc) 
 
# 배운 것 

## 이슈 

검색창에 단어를 입력하면 검색 자동완성 이벤트가 발동해 검색어 목록을 보여준다. 리스팅 된 검색어로 검색을 하기 위해 클릭을 했는데 click 이벤트가 발동하지 않는다. 어떻게 된걸까? mouse event와 blur 이벤트의 순서에 대해 이해하면 원인을 알 수 있다. mousedown, mouseup, click & blur 이벤트 순서에 대해 알아보자.  

![blur-mousedown](https://user-images.githubusercontent.com/35516239/61924477-f450a780-afa2-11e9-9e90-eb71d2331b6e.gif)


## mousedown, mouseup, click & blur 이벤트 순서

[이해를 위한 코드펜 링크](https://codepen.io/godotwait/pen/xvVMYp/)를 먼저 참고하자

1. 사용자 마우스를 누르면서 &rarr; `mousedown event` 발동
2. (특정 영역에 `focus` 된 요소가 있다면 해당 요소의 `blur`(`foucsout`둘의 차이는 `bubbling` 에 있다) `event` 발동
3. 마우스에서 손가락이 떨어지며   &rarr; `mouseup event` 발동
4. `click event` 발동 

## 문제 원인 파악 및 해결

`click event`가 일어나기 전에 `blur event`가 먼저 발생하여 `click event`가 발생하기 전에  검색어 링크가 `dom` 에서 삭제된다. `click event`  대신에 `mousedown event`를 활용하면 `blur event` 발동과 무관하게 원하는 로직을 사용할 수 있다.

Commit:
feat 검색이 완료(click, enter)됬을 때, result영역 지우기

- 이외 변수명 수정
- 조건 함수로 변경

Commit:
fix: hover가  clear가 되는 경우  검색결과가 없을경우 이벤트 처리

Commit:
refector: 파일 구조 변경, ResultViewRenderer 분리

Commit:
feat: getData 함수 및 api 요청에 대한 응답 추가

Commit:
refect: autoSuggestion, carousel 병합 및 main 폴더 생성

Commit:
chore: blur 이벤트 주석 처리 변경

Commit:
typo : tpl-src-data.json 내부 type 수정

Commit:
fix: insertAdjacentHTML 인자 변경 afterbegin

crongro commented:
버그 원인을 정확히 찾아서 해결했군요!
회사에서 하는 것과 매우 비슷하네요.
🥇

crongro commented on Main/main.js:
```diff
@@ -0,0 +1,18 @@
+
+window.addEventListener('DOMContentLoaded', () => {
+  renderHTML({
+    currentVersion: 1,
```
currentVersion 이 뭐에요?

p-iknow replied:
cache 정책 사용으로 필요한 변수인데,  업데이트가 있을 경우 배포시에 버전을 올려 localstorage에 저장된 버전번호와 비교해서 버전이 다르면 캐쉬가 아닌 다시 데이터를 요청해서 마크업을 합니다.

crongro commented on Main/main.js:
```diff
@@ -0,0 +1,18 @@
+    currentVersion: 1,
+    url: '../static/tpl-src-data.json',
+    templateFunc: mainTpl
+  }).then(_ => {
```
아래에서 이미 잘 했지만 이부분도 async 함수를 써서 then메서드에서 벗어날 수 있을 듯 하군요.

crongro commented on PLib/render.js:
```diff
@@ -0,0 +1,21 @@
+    localStorage.setItem('version', currentVersion);
+    return data;
+  }
+  const cache = await localStorage.getItem('data');
```
localStorage가 비동기코드인가요? await가 필요한건가 확인해보세요.

p-iknow replied:
기존 코드가 localStorage.getItem의 반환값을 promise로 감싸는 코드였는데, async 내부에서는 그럴필요가 없었네요. 어차피 async 함수는 thenable한 프로미스를 리턴하기 때문이죠 ㅠㅠ 수정하겠습니다.

crongro commented on PLib/render.js:
```diff
@@ -0,0 +1,21 @@
+  return JSON.parse(cache);
+};
+const renderHTML = ({ currentVersion, url, templateFunc }) => {
+  const body = document.querySelector('body');
```
body는 좀 특별하게도, document.body 로 접근이 된다고 합니다.

p-iknow replied:
참고하겠습니다.

crongro commented on autoSuggestion/src/component/Input/InputView.js:
```diff
@@ -0,0 +1,61 @@
+    this.onSelect.classList.add(onSelect);
+  }
+
+  keyboardNavigate(resultEl, direction) {
```
클린코드 관점에서 전 이코드를 개선할 거 같아요.

if 아래 if 아래 삼항연산자(조건문이죠) ..

를 개선하면 좋겠네요. 
어렵다면 subroutine을 만드는 것이라도 해야할 듯.

p-iknow replied:
사실 저도 찝찝한 코드였는데, 나중으로 미루다가 여기까지 왔네요. 수정하겠습니다.

crongro commented on autoSuggestion/src/component/Result/ResultView.js:
```diff
@@ -0,0 +1,41 @@
+    this.resultEl.innerHTML = '';
+    this.resultEl.style.display = 'block';
+    let template;
+    if (!dataSrc) {
```
개인적으로는 부정보다 긍정조건이 먼저 오는게 읽기편하던데, 본인은 어떠신가유?

p-iknow replied:
저도 긍정조건이 더 편합니다. if문 중첩을 줄이기 위해 부정조건에서 return 하는 습관 때문에 이렇게 적게 됬네요.

crongro commented on autoSuggestion/src/component/Result/ResultViewRenderer.js:
```diff
@@ -0,0 +1,46 @@
+  suggestionTemplate(query, suggesions, { resultItemHighlighted, resultItem }) {
+    const pattern = new RegExp(`${query}`, 'i');
+    return suggesions.reduce((prev, curr) => {
+      const match = pattern.exec(curr);
```
정규표현식을 좀 엉뚱한곳에 썼다고 전 보여지는데요.
정규표현식의 활용이유와 이렇게 표현해서 얻어지는 이득이 무엇인가요?

p-iknow replied:
대소문자와 무관하게 highlight 을 적용시키기 위해 썻습니다.  
ex) "iphone"(query 소문자) => "Iphone"(source  I 대문자로 시작한 경우) 

replace 메소드를 좀 더 검색해보니, `string as parameter` 를 활용하여 replace 인자로 pattern 을 넣고 매칭되는 문자열을 `$&(일치하는 문자열)` 로 받아서 활용하는 편이 더 깔끔하네요.  

`const match = pattern.exec(curr);` 이 부분은 불필요하므로 수정하겠습니다.
```js
  suggestionTemplate(query, suggesions, { resultItemHighlighted, resultItem }) {
    const pattern = new RegExp(`${query}`, 'i');
    return suggesions.reduce((prev, curr) => {
      const hilghtedSuggestion = curr.replace(
        pattern,
        `<span class=${resultItemHighlighted}>$&</span>`
      );
      return `${prev}<a href="#" class="${resultItem}" data-value="${curr}"><li>${hilghtedSuggestion}</li></a>`;
    }, '');
  },
```

crongro replied:
좋습니다.

crongro reviewed (changes requested):
버그 수정 잘 했고요. 
이벤트 순서와 같이 애매한 것들이 있죠. 표준으로 정의된 순서가 없을 거 같고요.
그런 부분은 브라우저 마다 구현방식이 조금씩 다를 수도 있습니다.

코드리뷰를 하면서 궁금증과 약간 수정하면 좋겠다는 부분이 보이네요.
수정도 한번하고 가고, 답변도 달아주세요~

Commit:
refect: keyboardNavigate 함수 내부의 if depth 최소화

- getNextOnSelectEl 함수를 submodule로 만듬

Commit:
refetc: then chaning -> async 함수로 변경

- getNextOnSelectEl 함수 typo 수정
- body 를 얻기 위해 document.querySelector() 했던 것 document.body로
변경

Commit:
refect: localStorage.getItem 앞에 await 삭제

p-iknow commented:
일부 수정 사항 반영과 질문에 답변 완료했습니다.

crongro commented on autoSuggestion/src/component/Input/InputView.js:
```diff
@@ -24,26 +24,27 @@ class InputView {
+    this.inputEl.value = this.onSelect.dataset.value;
+  }
+
+  getNextOnSelectEl(direction, firstNode, lastNode) {
```
조건문이 좀 줄어서 보기 좋은거 같아요.

crongro commented on autoSuggestion/src/js/config.js:
```diff
@@ -7,6 +7,5 @@ export default {
   throttleDelay: 60,
-  srcUrl:
-    'https://h3rb9c0ugl.execute-api.ap-northeast-2.amazonaws.com/develop/amazon_autocomplete?query='
+  srcUrl: 'http://127.0.0.1:3000/auto-suggestion?maxresult=10&prefix='
```
configuration을 이렇게 js코드에 넣기도하고, 
사실은 build과정이 생기게 되면, build할때 설정파일을 읽어들여서 소스코드에 추가해주는 식으로 개발하게 되겠죠.
(webpack에서 config파일을 읽어서 코드에 반영해주는 방식)

crongro reviewed (approved):
전에도 말했지만 테스트 가능한 코드일까? 
항상 물어보고 개발하시면 좋은코드, 좋은함수가 만들어질겁니다.
(왠지 테스트코드 구현하는 거 좋아하실거 같은데...)
